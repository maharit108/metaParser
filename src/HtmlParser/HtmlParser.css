.main_Wrapper {
  display: flex;
  flex-direction: row;
}

.left_inputSection {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100vh;
  width: 50vw;
}

.errTxt {
  font-size: 13px;
  color: rgb(202, 10, 10);
  width: 80%;
  max-width: 450px;
  min-width: 280px;
  margin: 0;
  text-align: left;
}

.right_displaySection {
  display: flex;
  align-items: flex-start;
  justify-content: center;
  width: 50vw;
  margin-top: 80px;
}
