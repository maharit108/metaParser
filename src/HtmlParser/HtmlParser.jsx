import React, { Component } from 'react'
import { TextField, Button, Divider, Paper, Table, TableContainer, TableCell, TableBody, TableHead, TableRow } from '@mui/material';
import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import SortIcon from '@mui/icons-material/Sort';

import './HtmlParser.css'


class HtmlParser extends Component {
  constructor() {
    super()
    this.state = {
      htmlStr: '',
      metaObjArr: [],
      errMsg: '',
      nameisAscending: false,
    }
  }
  
  getMetaContents = (el) => {
    if (el.children.length === 0) {
      let metaObj = {}
      let name = ''

      // for EPUB2 when metadata uses meta tag's name and content attributes
      if (el.getAttribute('name')) {
        name = el.getAttribute('name')
        if (name.includes('a11y:')) {
          metaObj = {
            name: name.replace('a11y:', ''),
            content: el.getAttribute('content')
          }
          this.setState(prevState => ({ metaObjArr: [...prevState.metaObjArr, metaObj]}))
        }
      }

      // for EPUB3 when metadata uses link and meta tags with property, rel and href attributes
      if (el.getAttribute('property') || el.getAttribute('rel')) {
        name = el.getAttribute('property') || el.getAttribute('rel')
        if(name.includes('a11y:')) {
          if (el.getAttribute('rel')) {
            metaObj = {
              name: name.replace('a11y:', ''),
              content: el.getAttribute('href')
            }
          } else {
            metaObj = {
              name: name.replace('a11y:', ''),
              content: el.nextSibling.textContent
            }
          }
          this.setState(prevState => ({ metaObjArr: [...prevState.metaObjArr, metaObj]}))
        }
      }
      return
    }

    // recursive check all nodes of given element
    for (let i=0; i < el.children.length; i++) {
      this.getMetaContents(el.children[i])
    }
  }

  /*
  // create a dummy element and add HTML string to it to use browser to validate if the string is valid HTML or not
  // if valid, use DOM methods to get the required data
  */
  onParse = () => {
    this.setState(prevState => (prevState.metaObjArr.length && {metaObjArr: []}))
    let el = document.createElement('div');
    el.innerHTML = this.state.htmlStr
    if (Boolean(this.state.htmlStr) && Boolean(el.children.length)) {
      this.setState({errMsg:''})
      this.getMetaContents(el)
    } else {
      this.setState({errMsg: 'Please provide valid HTML', htmlStr: ''})
    }
  }

  sortFunc = (arr, sortBy, sortType) => {
    arr.sort(function(item1, item2) {
      var checkItem1 = item1[sortBy].toUpperCase(); // ignore upper and lowercase
      var checkItem2 = item2[sortBy].toUpperCase(); // ignore upper and lowercase
      if (checkItem1 < checkItem2) {
        return sortType ? -1 : 1;
      }
      if (checkItem1 > checkItem2) {
        return sortType ? 1 : -1;
      }
      // names must be equal
      return 0;
    });
    this.setState({ 
      metaObjArr: arr, 
      nameisAscending: sortBy === 'name' ? !this.state.nameisAscending : this.state.nameisAscending, 
     })
  }

  render () {
    return (  
      <div className="main_Wrapper">
        <div className="left_inputSection">
          <TextField
            label="HTML Metadata"
            multiline
            rows={10}
            variant="outlined"
            sx={{ width: '80%', maxWidth: '450px', minWidth: '280px'}}
            value={this.state.htmlStr} 
            onChange={(e) => this.setState({htmlStr: e.target.value})}
          />
          {this.state.errMsg && (<p className="errTxt">{this.state.errMsg}</p>)}
          <Button sx={{ width: '80%', maxWidth: '450px', minWidth: '280px', margin: '20px auto'}}
                  variant="contained" 
                  onClick={this.onParse}>
            Parse HTML
          </Button>
        </div>

        <Divider orientation="vertical" variant="middle" flexItem />
        
        <div className="right_displaySection">
          <Paper sx={{width: '80%', minWidth: '350px', maxWidth: '600px' }}>
            <TableContainer>
              <Table sx={{ width: '80%', minWidth: '350px', maxWidth: '600px' }} aria-label="Accessibility Meta Name and Contents">
                <TableHead>
                  <TableRow>
                    <TableCell onClick={() => this.sortFunc(this.state.metaObjArr, 'name', this.state.nameisAscending)}><SortIcon />Name</TableCell>
                    <TableCell >Content</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {this.state.metaObjArr.map((metaObj, idx) => (
                    <TableRow
                      key={idx}
                      sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
                    >
                      <TableCell align="left">{metaObj.name}</TableCell>
                      <TableCell align="left">{metaObj.content}</TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
            </TableContainer>
          </Paper>
        </div>
        
      </div>
    )
  }
}

export default HtmlParser
