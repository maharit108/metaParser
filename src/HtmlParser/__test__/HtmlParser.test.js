import React from 'react';
import ReactDOM from 'react-dom';
import {cleanup, fireEvent, render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';

import HtmlParser from '../HtmlParser';

afterEach(cleanup);

it('should render HtmlParser withour crash', () => {
  const div = document.createElement("div");
  ReactDOM.render(<HtmlParser />, div)
})

it('should match snapshot', () => {
  const { asFragment }= render(<HtmlParser />);
  expect(asFragment()).toMatchSnapshot();
})

it('should show error if input is not valid HTML string', () => {
  render(<HtmlParser />)
  const textArea = screen.getByTestId('input')
  expect(textArea).toBeInTheDocument()

  // fireEvent.change(textArea, {target: { value: 'apple' }})
  userEvent.type(textArea, 'apple')

  fireEvent.click(screen.getByRole('button'))

  const err = screen.getByTestId('errTxt')
  // expect(textArea.innerHTML).toMatch('apple')
  expect(err.innerHTML).toMatch('Please provide valid HTML')
})

it('should show error if input is empty', () => {
  render(<HtmlParser />)
  fireEvent.click(screen.getByRole('button'))
  const err = screen.getByTestId('errTxt')
  expect(err.innerHTML).toMatch('Please provide valid HTML')
})

it('should show name and content values parsed from HTML if valid HTML is provided', () => {
  render(<HtmlParser />)
  userEvent.type(screen.getByTestId('input'), `<meta property="a11y:certifiedBy">Dewey, Checkett and Howe</meta>`)
  fireEvent.click(screen.getByRole('button'))
  const name = screen.getByTestId('meta_name')
  expect(name.innerHTML).toMatch('certifiedBy')
})


