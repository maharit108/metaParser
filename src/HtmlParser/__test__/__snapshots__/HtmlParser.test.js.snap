// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="main_Wrapper"
  >
    <div
      class="left_inputSection"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1q9r6s8-MuiFormControl-root-MuiTextField-root"
        data-testid="input"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-2"
          id="mui-2-label"
        >
          HTML Metadata
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-2"
            rows="10"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-173wfxe"
            >
              <span>
                HTML Metadata
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-9x0c2x-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Parse HTML
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-middle MuiDivider-vertical MuiDivider-flexItem css-16bghm4-MuiDivider-root"
    />
    <div
      class="right_displaySection"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1pgpu5u-MuiPaper-root"
      >
        <div
          class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
        >
          <table
            aria-label="Accessibility Meta Name and Contents"
            class="MuiTable-root css-npujbu-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="SortIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 18h6v-2H3v2zM3 6v2h18V6H3zm0 7h12v-2H3v2z"
                    />
                  </svg>
                  Name
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Content
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
